packages/expo-file-system/android/src/main/java/expo/modules/filesystem/FileSystemModule.kt

```diff
@@ -0,0 +1,1356 @@
+import org.apache.commons.io.FileUtils
+import org.apache.commons.io.IOUtils
+import java.io.BufferedInputStream
+import java.io.ByteArrayOutputStream
+import java.io.File
+import java.io.FileInputStream
+import java.io.FileNotFoundException
+import java.io.FileOutputStream
+import java.io.IOException
+import java.io.InputStream
+import java.io.OutputStream
+import java.io.OutputStreamWriter
+import java.lang.ClassCastException
+import java.lang.Exception
+import java.lang.NullPointerException
+import java.math.BigInteger
+import java.net.CookieHandler
+import java.net.URLConnection
+import java.util.*
+import java.util.concurrent.TimeUnit
+import kotlin.math.pow
+
+private const val NAME = "ExponentFileSystem"
+private val TAG = FileSystemModule::class.java.simpleName
+private const val EXDownloadProgressEventName = "expo-file-system.downloadProgress"
+private const val EXUploadProgressEventName = "expo-file-system.uploadProgress"
+private const val MIN_EVENT_DT_MS: Long = 100
+private const val HEADER_KEY = "headers"
+private const val DIR_PERMISSIONS_REQUEST_CODE = 5394
+
+class FileSystemModule(
+  context: Context,
+  private val moduleRegistryDelegate: ModuleRegistryDelegate = ModuleRegistryDelegate()
+) : ExportedModule(context), ActivityEventListener {
+
+  private inline fun <reified T> moduleRegistry() =
+    moduleRegistryDelegate.getFromModuleRegistry<T>()
+
+  private val uIManager: UIManager by moduleRegistry()
+  private var client: OkHttpClient? = null
+  private var dirPermissionsRequest: Promise? = null
+  private val taskHandlers: MutableMap<String, TaskHandler> = HashMap()
+
+  private enum class UploadType(private val value: Int) {
+    INVALID(-1), BINARY_CONTENT(0), MULTIPART(1);
+
+    companion object {
+      fun fromInt(value: Int): UploadType =
+        values().find { it.value == value } ?: INVALID
+    }
+  }
+
+  override fun onCreate(moduleRegistry: ModuleRegistry) {
+    moduleRegistryDelegate.onCreate(moduleRegistry)
+  }
+
+  override fun getName() = NAME
+
+  override fun getConstants(): Map<String, Any> {
+    return mapOf(
+      "documentDirectory" to Uri.fromFile(context.filesDir).toString() + "/",
+      "cacheDirectory" to Uri.fromFile(context.cacheDir).toString() + "/",
+      "bundleDirectory" to "asset:///",
+    )
+  }
+
+  private fun uriToFile(uri: Uri): File {
+    return File(uri.path)
+  }
+
+  @Throws(IOException::class)
+  private fun checkIfFileExists(uri: Uri) {
+    val file = uriToFile(uri)
+    if (!file.exists()) {
+      throw IOException("Directory for " + file.path + " doesn't exist.")
+    }
+  }
+
+  @Throws(IOException::class)
+  private fun checkIfFileDirExists(uri: Uri) {
+    val file = uriToFile(uri)
+    val dir = file.parentFile
+    if (dir == null || !dir.exists()) {
+      throw IOException("Directory for ${file.path} doesn't exist. Please make sure directory '${file.parent}' exists before calling downloadAsync.")
+    }
+  }
```

Comment:
Those methods can be converted to be an extension of the Uri class. Maybe it will make our code more readable?